muxers/mplex/src/io.rs

```diff
@@ -0,0 +1,733 @@
+                                e.remove();
+                                let below_limit = self.open_substreams.len() == self.config.max_substreams - 1;
+                                if below_limit {
+                                    ArcWake::wake_by_ref(&self.notifier_open);
+                                }
+                            },
+                            SubstreamState::Open => {
+                                debug!("Substream half-closed by remote: {:?}", stream_id);
+                                e.insert(SubstreamState::RecvClosed);
+                            },
+                        }
+                    }
+                }
+                Frame::Reset { stream_id } => {
+                    if let Some(state) = self.open_substreams.remove(&stream_id.into_local()) {
+                        debug!("Substream {:?} in state {:?} reset by remote", stream_id, state);
+                        let below_limit = self.open_substreams.len() == self.config.max_substreams - 1;
+                        if below_limit {
+                            ArcWake::wake_by_ref(&self.notifier_open);
+                        }
+                    } else {
+                        trace!("Received reset for unknown stream: {:?}", stream_id);
+                    }
+                }
+                _ => ()
+            }
+
+            if let Some(frame) = select(&frame) {
+                // The caller is interested in this frame.
+                return Poll::Ready(Ok(Some(frame)));
+            } else {
+                // The caller is not interested in this frame, so we need to buffer
+                // it and wake those tasks who are interested.
+                let id = frame.local_id();
+                if self.can_read(&id) || frame.is_open() {
+                    trace!("Buffering frame: {:?} (total: {})", frame, self.buffer.len() + 1);
+                    self.buffer.push(frame);
+                    self.notifier_read.wake_by_id(id);
+                } else if frame.is_data() {
+                    debug!("Dropping {:?} for closed or unknown substream: {:?}", frame, id);
+                }
+            }
+        }
+    }
+
+    /// Generates the next outbound stream ID.
+    fn next_outbound_stream_id(&mut self) -> LocalStreamId {
+        let id = self.next_outbound_stream_id;
+        self.next_outbound_stream_id = self.next_outbound_stream_id.next();
+        id
+    }
+
+    /// Checks whether a substream is open for reading.
+    fn can_read(&self, id: &LocalStreamId) -> bool {
+        match self.open_substreams.get(id) {
+            Some(SubstreamState::Open) | Some(SubstreamState::SendClosed) => true,
+            _ => false,
+        }
+    }
+
+    /// Sends pending frames, without flushing.
+    fn send_pending_frames(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        while let Some(frame) = self.pending_frames.pop() {
```

Comment:
I don't know if it was already like this before, but I find it strange that `pending_frames` is a FILO queue here. I would expect entries to be sent in the order they have been pushed.

Reply:
Yes, that is certainly better.

> I don't know if it was already like this before ...

There were no `pending_frames` before, since they were introduced primarily to send close/reset frames for dropped substreams as necessary, i.e. #313.